github_actions/lib/dependabot/github_actions/update_checker.rb

```diff
@@ -122,14 +119,21 @@ def updated_source
           return dependency_source_details.merge(ref: new_tag.fetch(:tag))
         end
 
+        latest_tag = git_commit_checker.local_tag_for_latest_version
+
         # Update the git commit if updating a pinned commit
         if git_commit_checker.pinned_ref_looks_like_commit_sha? &&
-           (latest_tag = git_commit_checker.local_tag_for_latest_version) &&
            git_commit_checker.branch_or_ref_in_release?(latest_tag[:version]) &&
            (latest_commit = latest_tag.fetch(:commit_sha)) != current_commit
           return dependency_source_details.merge(ref: latest_commit)
         end
 
+        # Update to the latest tag if pinned to a commit that has diverged
```

Comment:
I'd still expect this to update to the commit sha of the latest version rather than switch to a version here. Maybe we just need to remove the `git_commit_checker.branch_or_ref_in_release?(latest_tag[:version])` above?

If we had v1 (sha: a) and latest version v2 (sha: b) I'd expect:
1. workflow referencing v1 = PR that updates to v2
2. workflow referencing sha:a = PR that updates to sha: b

The #2 case isn't working with diverged commits now but it should.

Reply:
In my opinion, there's no easy way of knowing what the user wants in this case. I think there's two use cases for direct sha references:

> I want to pick up a bug fix or feature not yet released, because I really need it, but once released I want to stay on stable versions.

In the above case, user will expect dependabot to bring her back to a regular version when possible.

> I love what this action is doing and I really want to be on top of the latest changes as soon as they hit the default branch.

In the above case, user will expect dependabot to keep using a commit reference and keep it up to date.

So, I guess the optimal solution would be to make it configurable, but probably too much of an edge case for that.

Reply:
In any case, what we are doing after this PR feels better because we keep the same style in the workflow, so whatever the end user expectation is, this is likely to be less surprising than what we were doing previously 👍.